### ISSUE TITLE
Make `.RenderString` render shortcodes

### ISSUE THREAD

**Issue by regisphilibert** [gohugoio/hugo#6703]
As [requested](https://github.com/gohugoio/hugo/issues/6666#issuecomment-568779670) I'm re-creating this issue.

Now that we have `.RenderString`. The problem mentioned in https://github.com/gohugoio/hugo/issues/6094#issuecomment-513116608 has been lifted and we should consider having `.RenderString` process shortcodes.

### Do we really need this?

When building a page with blocks of data (CMS Blocks,  Page Builders and so on), most of those are saved as Front Matter. It then falls on the developer to range on those blocks and apply the proper data processing/partial loading etc...

In the context of a very nice to have page builder, Hugo shortcodes are currently unusable by the editor and any search-and-replace workaround cooked up by developers (been there) will never match the reliability and usability of Hugo's shortcode API.

### Will it break code.

The `.RenderString` method is fairly new and I don't see anyone including shortcodes in there with the expectation that they would not be processed. So I think: No

**Comment by aaronpowell** [comment]
I just came across `.RenderString` as a way to render some markdown that's in the front matter of a page and was surprised it didn't handle shortcodes (then I read the docs and saw it stated).

I see value in it handling that, if not, an alternative like `shortcodify` like was mentioned in the parent issue.

**Comment by mlake** [comment]
also seeing a need for this..

**Comment by bep** [comment]
Filling up my notification screen with "I need this, too" does not help.

**Comment by regisphilibert** [comment]
Hello @mlake, usually a "thumb up" emoji on the issue's description is enough to notify of your interest in this issue being treated. You'll get the same effect without the noise. Thanks a ton!

**Comment by NorseGaud** [comment]
This is a huge need for us. Partials allow for a lot of DRYness in our docs site and not being able to include markdown with other shortcodes in it (like relrefs) is painful.

**Comment by bric3** [comment]
Hi I stumbled on this issue, and I wonder if the _new_ `.RenderString` that could process shortcodes could help me in the below case.

----

I have a mardown table that I'd like to wrap in a div, and I'd have shortcodes to highlight / mark some text in this table.

```
{{< wrapTable >}}

| Memory Type | # of Segments | Used Memory(bytes) | Used Memory(%) | Free Memory(bytes) | Free Memory(%) | Total Memory(bytes) |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| Internal | 13 | 1 191 740 | {{< mark >}}**88,4**{{< /mark >}} | 156 452 | 11,6 | 1 348 192 |
| Object (reversed) | 1 | 402 653 184 | {{< mark >}}**100**{{< /mark >}} | 0 | 0 | {{< mark >}}**402 653 184**{{< /mark >}} |
| Class | 9 735 | 228 637 344 | {{< mark >}}**90,96**{{< /mark >}} | 22 724 116 | 9,04 | {{< mark >}}**251 361 460**{{< /mark >}} |
| JIT Code Cache | 5 | 41 943 040 | {{< mark >}}**100**{{< /mark >}} | 0 | 0 | 41 943 040 |
| JIT Data Cache | 3 | 17 018 496 | 67,63 | 8 147 328 | 32,37 | 25 165 824 |
| Overall | 9 757 | 691 443 804 | 95,71 | 31 027 896 | 4,29 | 722 471 700 |

{{< /wrapTable />}}
```

My shortcodes are 

* `wrapTable.html`
```
<div class="table-wrapper">
    <!-- {{ .Inner | markdownify }} -->
</div>
```
* `mark.html`
```
<mark>{{ .Inner | markdownify }}</mark>
```

However with this code the table is broken as soon as there's the `{{< mark >}}` shortcode.

I've tried using `$.Page.RenderString`  instead of `mardownify` but this didn't actually change anything. So for now I'm using a hacky way with two shortcodes for the table wrapper.

* `openWrapTable` that output the `<div>`
* `closeWrapTable` that output `</div>`

**Comment by nstickney** [comment]
Found this; my use case (which I've not yet seen mentioned in any of the various issue discussions so far, apologies if I missed it) is shortcodes with [Remark.js](https://github.com/gnab/remark). Typical implementation uses `.RawContent`, which of course means shortcodes go un-rendered. Being able to do `.RawContent | shortcodify`, or `.RawContent | shortcodify .` to [have the `.Page` context](https://github.com/gohugoio/hugo/issues/6094) if necessary would be excellent.

**Comment by zabatonni** [comment]
I'm really missing this feature. I need to process youtube links and "ref" links from frontmatter blocks. Now I'm using really wild regex...

**Comment by kevinkenan** [comment]
Hmmm....this seems to be working in `v0.81.0+extended`.

**Comment by zabatonni** [comment]
> ~Hmmm....this seems to be working in `v0.81.0+extended`.~
> 
> Never mind...I was looking at the wrong case in my tests.

Not working, confirmed on v0.82.0

**Comment by jmooring** [comment]
Test case:

```
git clone --single-branch -b hugo-github-issue-6703 https://github.com/jmooring/hugo-testing hugo-github-issue-6703
cd hugo-github-issue-6703
hugo server
```

**Comment by gumaerc** [comment]
I'm working on the next generation of MIT Open CourseWare and we're using Hugo to build course websites.  This functionality would be very useful for us as we use a JSON data template to store a lot of information about the course.  We have at least one field in there right now where we store markdown text and we need to be able to process shortcodes through either `markdownify` or `RenderString`.

**Comment by gumaerc** [comment]
Another place I've noticed that this being a problem recently is when you have an outer and inner shortcode and the outer shortcode is being called with the `%` delimiters.  For example:

`outer.html`:
```
<div>
{{ .Inner }}
</div>
```

`mid.html`:
```
<p>
{{ .Inner }}
</p>
```

`inner.html`:
```
{{ .Get 0 }}
```

`test.md`:
```
{{< outer >}}
{{% mid %}}
{{< inner "test" >}}
{{% /mid %}}
{{< outer />}}
```

**Comment by bep** [comment]
@regisphilibert would you expect `$page.HasShortcode "foo"` to return true if `foo` was used only in `$page.RenderString`?

**Comment by regisphilibert** [comment]
Yes! I would think so. `.HasShortcode` is usually used to load a dedicated script or style in the page in case the shortcode is used. I understand it could have value to know if given shortcode is used through `.RenderString` or `.Content` but can't cope with not knowing either or as long as I know the shortcode is used on the page.
